### ISSUE TITLE
Strategy for associating styles to components rather than at the parent component level?

### ISSUE THREAD

**Issue by kprokopczyk** [wix-incubator/mjml-react#57]
I have a base template that i pass my components to and render them inside `MjmlBody`:
```
const BaseEmailTemplate = ({ children }) => {
    const styles = readFileSync(resolve(__dirname, '../../assets/styles.css')).toString();
    return (
        <Mjml>
            <MjmlHead>
                <MjmlStyle inline>{styles}</MjmlStyle>
            </MjmlHead>
            <MjmlBody>
                {children}
            </MjmlBody>
        </Mjml>
    );
};
```
So I'm reading in a single css file and rendering the styles in `MjmlHead`, much like your example. But maintaining a giant `styles.css` is not ideal in terms of maintainability or scalability - each of my templates includes ALL the styles, even for components the template does not include. 

I'm wondering if you have any strategies for moving css to the component level so that only the styles associated with the `children` are rendered in head.  

Thanks for any tips.

**Comment by IanEdington** [comment]
> So I'm reading in css files and rendering the styles in MjmlHead, much like your example.

What example are you referring to? We have been talking about how to solve this problem as well and are planning on taking the approach that `styled-components` uses for render to string... just haven't had a chance to implement yet

**Comment by kprokopczyk** [comment]
The example I'm referring to is just this, where a file is read in: https://github.com/wix-incubator/mjml-react-example/blob/master/src/email.js#L18

A post processor feels like a good stopgap but it still doesn't allow me to couple styles with components, so I'll probably keep looking around for a good solution. Thanks!

**Comment by IanEdington** [comment]
yeah agreed. Same reason we aren't doing it

**Comment by IanEdington** [comment]
@kprokopczyk we've started using https://github.com/styled-components/styled-components with mjml-react. 

This is our "react to mjml" script:
```
import React from "react";
import ReactDOMServer from "react-dom/server";
import { ServerStyleSheet } from "styled-components";

function getStyledComponentsStaticCss(sheet: any): string {
  const tag = sheet.getTag();

  return [...Array(tag.length).keys()].map((group) => tag.getGroup(group)).join("");
}

export function getMjmlFromReact(element: React.ReactNode) {
  const sheet = new ServerStyleSheet();
  const elementWithCollectedStyles = sheet.collectStyles(element);
  const mjmlWithoutStyledComponentsCss = ReactDOMServer.renderToStaticMarkup(elementWithCollectedStyles);
  const css = getStyledComponentsStaticCss(sheet.instance);
  return mjmlWithoutStyledComponentsCss.replace(/\/\* inject css here \*\//g, css);
}
```

How to use styled components:
```
const StyledMjmlText = styled(({className, ...props}) => <MjmlText {...props} cssClass={className} />)`
  height: 250px;
  border: 1px solid red !important;
`;

export const TestingStyles: Story = () => (
  <Mjml>
    <MjmlHead>
      <MjmlStyle>
        {css`
          /* inject css here */
        `}
      </MjmlStyle>
    </MjmlHead>
    <MjmlBody>
      <MjmlSection>
        <MjmlColumn>
          <StyledMjmlText>Hello 250px high typography</StyledMjmlText>
        </MjmlColumn>
      </MjmlSection>
    </MjmlBody>
  </Mjml>
);
```

This works with deeply nested components. Unfortunately it requires `className` => `css-class` which is a little cumbersome.

Based on this repo basically no support. I'm thinking about just implementing my own version with a bunch more functionality like `className` => `css-class`

**Comment by gajus** [comment]
@IanEdington Did you end up developing / open-sourcing anything to help with styled-components adoption?

**Comment by gajus** [comment]
Tried your code and I am getting:

> TypeError: sheet.getTag is not a function

What React / `styled-components` version are you using?

**Comment by IanEdington** [comment]
@gajus not yet but I could... let me get back to you

**Comment by gajus** [comment]
I would appreciate that. I spent couple of hours playing with styled-components and Mjml and couldn't make it work.

**Comment by IanEdington** [comment]
@gajus I got the go ahead to open source a couple of our mjml tools. I should be able to get them up early next week

**Comment by gajus** [comment]
> @gajus I got the go ahead to open source a couple of our mjml tools. I should be able to get them up early next week

Please ping in this thread if / when you share those tools. 🙏

**Comment by IanEdington** [comment]
Sorry for the delay on this. I'm trying to prioritize this work.

**Comment by gajus** [comment]
> Sorry for the delay on this. I'm trying to prioritize this work.

Any chance you could share whatever MVP you've got, even it isn't polished?

**Comment by matepaiva** [comment]
@IanEdington please let us know when you share this! <3 Thank you!

**Comment by gajus** [comment]
@IanEdington Any luck?

**Comment by IanEdington** [comment]
Hey all, sorry for leaving you hanging. We have support for this in https://github.com/Faire/mjml-react version 3+

This issue has a short overview of how we use style-components https://github.com/Faire/mjml-react/issues/70

TLDR: We've added `className` -> `cssClass` so any react tool that uses className can be used with mjml-react

**Comment by matepaiva** [comment]
Amazing! Thanks @IanEdington!
